### ISSUE TITLE
DC-LLP-018: Tiny LLM router — multi-constraint reasoning (Tier 3)

### ISSUE THREAD

**Issue by radoxtech** [radoxtech/diricode#413]
## DC-LLP-018 — Tiny LLM Router (Tier 3)

**Epic**: #3 (Provider Router — Transport Layer)
**Package**: `@diricode/dirirouter` (`packages/dirirouter/`)
**Iteration**: v2.0
**ADR**: [ADR-049](docs/adr/adr-049-llm-picker.md), [ADR-055](docs/adr/adr-055-diri-router-unified-package.md)
**Previously**: LLM Picker epic #389 (closed) → Sub-Epic #393 (closed)

---

### Goal

Implement the Tier 3 router using a tiny LLM (e.g., Qwen-0.5B or Phi-2) running locally via ONNX Runtime. This tier handles complex, multi-constraint reasoning that Tier 1 heuristics and Tier 2 classifiers cannot resolve — such as balancing cost vs. quality vs. speed for ambiguous prompts.

### Why This Matters

Some routing decisions require genuine reasoning: "I need a model that can handle 100K context, costs under $0.01/1K tokens, and supports function calling" involves multiple constraints that can't be resolved by pattern matching or embeddings alone. Tier 3 is the "brain" of the Picker cascade — expensive but accurate, only invoked when simpler tiers fail.

#### Model Selection
- Primary: `Qwen2.5-0.5B` (500M params, ~200ms inference on CPU)
- Alternative: `Phi-2` (2.7B params, ~500ms — only if 0.5B insufficient)
- Format: ONNX quantized (INT8) for acceptable CPU performance
- Model files: `packages/dirirouter/models/qwen-0.5b-q8.onnx` (git-lfs)

#### Reasoning Prompt Template
```
You are a model router. Given the user's task and available models, select the best model.

Task: {prompt_summary}
Constraints: {extracted_constraints}
Available models: {model_cards_json}

Respond with JSON: { "model_id": "...", "reasoning": "...", "confidence": 0.0-1.0 }
```

#### Constraint Extraction
- Parse user prompt for explicit constraints (cost mentions, speed requirements, context length needs)
- Combine with implicit signals from Tier 2 classification
- Build `ConstraintSet`: `{ maxCost?, minQuality?, maxLatency?, requiredCapabilities?, contextLength? }`

#### Inference Pipeline
1. Extract constraints from prompt + Tier 2 output
2. Build reasoning prompt with constraint set + filtered model cards
3. Run ONNX inference (quantized tiny LLM)
4. Parse JSON output, validate against model registry
5. If parsing fails or confidence < 0.5: fall back to Tier 1 best-effort selection

### Acceptance Criteria

- [ ] ONNX quantized model loads and runs inference successfully
- [ ] Reasoning latency < 300ms on standard CPU
- [ ] Deterministic output for identical inputs (temperature = 0, no sampling)
- [ ] Correct constraint evaluation: when given "cheap and fast", selects appropriately vs "high quality reasoning"
- [ ] Graceful fallback to Tier 1 on model load failure, timeout (>500ms), or unparseable output
- [ ] JSON output schema validated before use (model_id must exist in registry)
- [ ] Memory footprint < 500MB for loaded quantized model
- [ ] Unit tests cover: multi-constraint resolution, fallback scenarios, determinism, timeout handling

### Architecture Notes

- Lives in `packages/dirirouter/src/picker/tiers/tier3-reasoning.ts`
- Constraint extractor: `packages/dirirouter/src/picker/constraints.ts`
- Shares ONNX Runtime with Tier 2 (single process, different models)
- Implements `PickerTier` interface: `reason(prompt, constraints, modelCards) → { modelId, reasoning, confidence, latency }`
- Model warming: load on first use, keep in memory until process exit

### References

- [ADR-049](docs/adr/adr-049-llm-picker.md) — LLM Picker cascade architecture
- [ADR-055](docs/adr/adr-055-diri-router-unified-package.md) — Unified diri-router package
- #412 — Tier 2 classifier (feeds classification + constraints into Tier 3)
- #327 — Local model metadata registry (provides model cards for reasoning)
- #422 — Auto re-training pipeline (can fine-tune this model)

### Dependencies

- Depends on: #412 (Tier 2 classification output), #327 (model metadata registry), ONNX Runtime
- Required by: #422 (re-training pipeline), #435 (integration with routing stack)

**Comment by radoxtech** [comment]
## Closing — Anti-pattern confirmed by research

After extensive web research on LLM routing systems, the TinyLLM Router approach (using an LLM to classify/route to other LLMs) is confirmed as an anti-pattern:

### Evidence:
1. **RouteLLM (Berkeley, ICLR 2025)**: Their benchmark shows BERT classifiers underperform Matrix Factorization for routing. A TinyLLM adds latency without accuracy gains.
2. **LiteLLM complexity_router**: Uses rule-based heuristics (<1ms, zero API calls) — outperforms LLM-based routing in production.
3. **vLLM Semantic Router (2026)**: Composable signal architecture, explicitly avoids LLM-in-the-loop for classification.
4. **Martian, Unify**: Neural scoring models, but NOT tiny generative LLMs — they use discriminative models.
5. **LLMRouterBench**: No benchmark winner uses generative LLM for routing.

### Technical problems with this approach:
- **Latency**: 200-500ms ONNX inference vs <1ms heuristics — 200-500x slower
- **Recursion risk**: LLM routing to LLMs creates dependency loops
- **Model management**: 500MB+ ONNX model in `git-lfs` for marginal accuracy gain
- **The `task.description` problem**: Current `DecisionRequest` has it as optional — TinyLLM has nothing to reason about when absent

### Replacement:
- **DC-DR-040** (Embedding Router): Zero-shot cosine similarity, <10ms, no training data needed
- **#412** (BERT Classifier): Moved to Phase 2, only after telemetry data collected via DC-DR-041

This issue's dependency chain (#412 → #413 → #422) is being restructured. See Epic #3.
